runtime/lua/vim/lsp.lua

```diff
@@ -16,24 +7,48 @@ local tbl_isempty, tbl_extend = vim.tbl_isempty, vim.tbl_extend
 local validate = vim.validate
 local if_nil = vim.F.if_nil
 
-local lsp = {
-  protocol = protocol,
-
-  handlers = default_handlers,
-
-  buf = require('vim.lsp.buf'),
-  diagnostic = require('vim.lsp.diagnostic'),
-  codelens = require('vim.lsp.codelens'),
-  inlay_hint = require('vim.lsp.inlay_hint'),
-  semantic_tokens = semantic_tokens,
-  util = util,
-
-  -- Allow raw RPC access.
-  rpc = lsp_rpc,
-
-  -- Export these directly from rpc.
-  rpc_response_error = lsp_rpc.rpc_response_error,
+local submodules = {
+  '_changetracking',
+  '_completion',
+  '_dynamic',
+  'buf',
+  'codelens',
+  'diagnostic',
+  'handlers',
+  'inlay_hint',
+  'log',
+  'protocol',
+  'rpc',
+  'semantic_tokens',
+  'tagfunc',
+  'util',
 }
+local dundarload_submodules = vim._dundarload_submodules('vim.lsp', submodules)
+local lsp = setmetatable({}, dundarload_submodules)
+
+lsp._changetracking = nil --- @module 'vim.lsp._changetracking'
+lsp._completion = nil --- @module 'vim.lsp._completion'
+lsp._dynamic = nil --- @module 'vim.lsp._dynamic'
+lsp.buf = nil --- @module 'vim.lsp.buf'
+lsp.codelens = nil --- @module 'vim.lsp.codelens'
+lsp.diagnostic = nil --- @module 'vim.lsp.diagnostic'
+lsp.handlers = nil --- @module 'vim.lsp.handlers'
+lsp.inlay_hint = nil --- @module 'vim.lsp.inlay_hint'
+lsp.log = nil --- @module 'vim.lsp.log'
+lsp.protocol = nil --- @module 'vim.lsp.protocol'
+lsp.rpc = nil --- @module 'vim.lsp.rpc'
+lsp.semantic_tokens = nil --- @module 'vim.lsp.semantic_tokens'
+lsp.tagfunc = nil --- @module 'vim.lsp.tagfunc'
+lsp.util = nil --- @module 'vim.lsp.util'
```

Comment:
What do you think if we mangled this to something like:

```lua
local lsp = {
    _changetracking = {} --- @module 'vim.lsp._changetracking'
    _completion = {} --- @module 'vim.lsp._completion'
    _dynamic = {} --- @module 'vim.lsp._dynamic'
    buf = {} --- @module 'vim.lsp.buf'
    codelens = {} --- @module 'vim.lsp.codelens'
    diagnostic = {} --- @module 'vim.lsp.diagnostic'
    handlers = {} --- @module 'vim.lsp.handlers'
    inlay_hint = {} --- @module 'vim.lsp.inlay_hint'
    log = {} --- @module 'vim.lsp.log'
    protocol = {} --- @module 'vim.lsp.protocol'
    rpc = {} --- @module 'vim.lsp.rpc'
    semantic_tokens = {} --- @module 'vim.lsp.semantic_tokens'
    tagfunc = {} --- @module 'vim.lsp.tagfunc'
    util = {} --- @module 'vim.lsp.util'
}

dundarload_submodules('vim.lsp', lsp)

locoal function dundarload_modules(root, mod)
    local submodules = vim.tbl_keys(mod)

    --- clear submodules
    for k in pairs(mod) do
        mod[k] = nil
    end

    setmetatable(mod, {
      __index = function(t, k)
        ---@diagnostic disable:no-unknown
        if vim.list_contains(submodules, k) then
          t[k] = require(string.format('%s.%s', root, k))
          return t[k]
        end
      end
    })
end
```

or is that going too far?